### PR TITLE
Refactor server api tests and dataset_manifest

### PR DESCRIPTION
**Pull request by zhiltsov-max**
### Motivation and context

Extracted from #5083 

- Fixed test data placement in tests
- Fixed cache cleaning in tests
- Refactored server API tests and dataset_manifest
- Added more tests

### Checklist

- [ ] I submit my changes into the `develop` branch
- [ ] I have added a description of my changes into the [CHANGELOG](https://github.com/opencv/cvat/blob/develop/CHANGELOG.md) file
- [ ] I have updated the documentation accordingly
- [ ] I have added tests to cover my changes
- [ ] I have linked related issues (see [GitHub docs](
  https://help.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword))
- [ ] I have increased versions of npm packages if it is necessary
  ([cvat-canvas](https://github.com/opencv/cvat/tree/develop/cvat-canvas#versioning),
  [cvat-core](https://github.com/opencv/cvat/tree/develop/cvat-core#versioning),
  [cvat-data](https://github.com/opencv/cvat/tree/develop/cvat-data#versioning) and
  [cvat-ui](https://github.com/opencv/cvat/tree/develop/cvat-ui#versioning))

### License

- [ ] I submit _my code changes_ under the same [MIT License](
  https://github.com/opencv/cvat/blob/develop/LICENSE) that covers the project.
  Feel free to contact the maintainers if that's a concern.

### CODE REVIEW

File: cvat/apps/engine/tests/utils.py [hunk 1]
```diff
@@ -2,14 +2,111 @@
...
+from PIL import Image
+from rest_framework.test import APIClient, APITestCase
+import av
+import numpy as np
 
 T = TypeVar('T')
 
+
+@contextmanager
+def disable_logging():
```
Comment (SpecLad) [inline cvat/apps/engine/tests/utils.py:23]: I would suggest to rename this to something like `log_disabler`, otherwise it's not clear that you're supposed to use it as a context manager.

File: cvat/apps/engine/task.py [hunk 2]
```diff
@@ -717,8 +717,8 @@ def _update_status(msg):
                         manifest.create()
                         _update_status('A manifest had been created')
 
-                        all_frames = len(manifest.reader)
-                        video_size = manifest.reader.resolution
+                        all_frames = len(manifest._reader)
+                        video_size = manifest._reader.resolution
```
Comment (SpecLad) [inline cvat/apps/engine/task.py:721]: This seems odd; why did you remove the `reader` property if it was used here?

File: cvat/apps/engine/tests/test_rest_api.py [hunk 3]
```diff
@@ -3,10 +3,9 @@
 #
 # SPDX-License-Identifier: MIT
 
-
+from contextlib import ExitStack
 import io
 import os
-import os.path as osp
```
[the remark this answers is not in the document]
Comment (zhiltsov-max) [reply cvat/apps/engine/tests/test_rest_api.py:9]: It was my standard for Datumaro. Sometimes I try to apply it in CVAT too, but in general, os.path is used throughout CVAT. I want to apply osp everywhere later, but now it just introduces inconsistency in some files.

File: cvat/apps/engine/task.py (hunk 2 above)
Comment (zhiltsov-max) [reply cvat/apps/engine/task.py:721]: Hmm, right. Thanks

File: cvat/apps/engine/tests/utils.py (hunk 1 above)
Comment (zhiltsov-max) [reply cvat/apps/engine/tests/utils.py:23]: How about `logging_disabled`? I think, `with logging_disabled():` reads nice.

File: cvat/apps/engine/tests/utils.py [hunk 4]
```diff
@@ -2,14 +2,111 @@
...
+        return self
+
+    def __exit__(self, exception_type, exception_value, traceback):
+        if self.user:
+            self.client.logout()
+
+
+class ApiTestBase(APITestCase):
+    def setUp(self):
+        self.client = APIClient()
```
[the remark this answers is not in the document]
Comment (zhiltsov-max) [reply cvat/apps/engine/tests/utils.py:51]: Added

File: cvat/apps/engine/tests/utils.py (hunk 1 above)
Comment (SpecLad) [reply cvat/apps/engine/tests/utils.py:23]: I don't know, that name is also ambiguous, as it suggests a boolean return value. I guess it's alright.

File: cvat/apps/engine/tests/test_rest_api.py [hunk 5]
```diff
@@ -3086,62 +3017,22 @@ def test_api_v2_tasks_id_export_no_auth(self):
...
-    f = BytesIO()
     gen = random.SystemRandom()
     width = gen.randint(100, 800)
     height = gen.randint(100, 800)
-    image = Image.new('RGB', size=(width, height))
-    image.save(f, 'jpeg')
-    f.name = filename
-    f.seek(0)
-
+    f = _generate_image_file(filename, size=(width, height))
```
[the remark this answers is not in the document]
Comment (zhiltsov-max) [reply cvat/apps/engine/tests/test_rest_api.py:3023]: In general, it's not a good idea to have randomness in tests, because it makes the tests harder to reproduce. In this case, I decided to factor out only the controlled-size version, because this random variant is used everywhere in this file. The extracted function can be used in other tests for other components, there are lots of duplicated code to remove.

File: cvat/apps/engine/tests/utils.py [hunk 6]
```diff
@@ -2,14 +2,111 @@
...
+        for cache in caches.all(initialized_only=True):
+            cache.clear()
+
+        return super().tearDown()
+
+
+def generate_image_file(filename, size=(100, 100)):
+    f = BytesIO()
+    image = Image.new('RGB', size=size)
+    image.save(f, 'jpeg')
```
[the remark this answers is not in the document]
Comment (zhiltsov-max) [reply cvat/apps/engine/tests/utils.py:69]: Added an assert

File: cvat/apps/engine/tests/test_rest_api.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (SpecLad) [reply cvat/apps/engine/tests/test_rest_api.py:3023]: Okay, that's fair enough, but I still object to the fact that both of these functions are named `generate_image_file`. Can you rename the randomized one so that it's clear it's a different function?

File: cvat/apps/engine/tests/test_rest_api.py [hunk 7]
```diff
@@ -30,16 +29,20 @@
...
 from cvat.apps.engine.models import (AttributeSpec, AttributeType, Data, Job,
     Project, Segment, StageChoice, StatusChoice, Task, Label, StorageMethodChoice,
     StorageChoice, DimensionType, SortingMethod)
 from cvat.apps.engine.media_extractors import ValidateDimension, sort
 from cvat.apps.engine.tests.utils import get_paginated_collection
 from utils.dataset_manifest import ImageManifestManager, VideoManifestManager
 
+from cvat.apps.engine.tests.utils import (ApiTestBase, ForceLogin, logging_disabled,
+    generate_video_file)
+from cvat.apps.engine.tests.utils import generate_image_file as _generate_image_file
```
Comment (SpecLad) [inline cvat/apps/engine/tests/test_rest_api.py:44]:
```suggestion
from cvat.apps.engine.tests.utils import (ApiTestBase, ForceLogin, logging_disabled,
    generate_image_file, generate_video_file)
```

**Review by SpecLad — APPROVED** [review]
